src/ProxyProtocol.cc

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include <algorithm>
+#include "parser/BinaryTokenizer.h"
+#include "parser/Tokenizer.h"
+#include "ProxyProtocol.h"
+#include "sbuf/StringConvert.h"
+#include "sbuf/Stream.h"
+#include "SquidConfig.h"
+#include "SquidString.h"
+#include "StrList.h"
+
+#if HAVE_SYS_SOCKET_H
+#include <sys/socket.h>
+#endif
+#if HAVE_NETINET_IN_H
+#include <netinet/in.h>
+#endif
+#if HAVE_NETINET_IP_H
+#include <netinet/ip.h>
+#endif
+
+/// magic octet prefix for PROXY protocol version 1
+static const SBuf Proxy1p0magic("PROXY ", 6);
+
+/// magic octet prefix for PROXY protocol version 2
+static const SBuf Proxy2p0magic("\x0D\x0A\x0D\x0A\x00\x0D\x0A\x51\x55\x49\x54\x0A", 12);
+
+ProxyProtocol::Message::Message(const char *ver, const uint16_t len, const uint8_t cmd)
+    : version_(ver), length_(len), command_(Two::CommandType(cmd)) {}
+
+SBuf
+ProxyProtocol::Message::getAll(const char sep) const
+{
+    SBuf result;
+    char ipBuf[MAX_IPSTRLEN];
+    result.appendf(":version: %s\r\n", version_);
+    result.appendf(":command: %d\r\n", command_);
+    auto srcAddr = srcIpAddr;
+    srcAddr.applyClientMask(Config.Addrs.client_netmask);
+    result.appendf(":src_addr: %s\r\n", srcAddr.toStr(ipBuf, MAX_IPSTRLEN));
+    result.appendf(":src_port: %d\r\n", srcIpAddr.port());
+    result.appendf(":dst_addr: %s\r\n", dstIpAddr.toStr(ipBuf, MAX_IPSTRLEN));
+    result.appendf(":dst_port: %d\r\n", dstIpAddr.port());
+    for (const auto &tlv: tlvs) {
+        result.appendf("%d: ", tlv.type);
+        result.append(tlv.value);
+        result.append("\r\n");
+    }
+    return result;
+}
+
+SBuf
+ProxyProtocol::Message::getValues(const uint32_t headerType, const char sep) const
+{
+    SBuf result;
+    char ipBuf[MAX_IPSTRLEN];
+
+    if (headerType == Two::PP2_EXTRA_TYPE_VERSION)
+        result.append(version_);
+    else if (headerType == Two::PP2_EXTRA_TYPE_COMMAND)
+        result.appendf("%d", command_);
+    else if (headerType == Two::PP2_EXTRA_TYPE_SRC_ADDR) {
+        auto logAddr = srcIpAddr;
+        (void)logAddr.applyClientMask(Config.Addrs.client_netmask);
+        result.appendf("%s", logAddr.toStr(ipBuf, MAX_IPSTRLEN));
```

Comment:
Where possible, use cheaper/simpler/C++-friendlier SBuf::append() instead of appendf(). Also, if ToSBuf() does what we want, you may want to use that instead and hide low-level toStr() parameters that way.